File: benchmarks/mandelbrot/mandelbrot.go
package main

import "fmt"

func mandelbrot(c_re, c_im float64, max_iter int32) int32 {
	z_re, z_im := 0.0, 0.0
	for i := int32(0); i < max_iter; i++ {
		z_re_sq := z_re * z_re
		z_im_sq := z_im * z_im
		if z_re_sq+z_im_sq > 4.0 {
			return i
		}
		z_im = 2.0*z_re*z_im + c_im
		z_re = z_re_sq - z_im_sq + c_re
	}
	return max_iter
}

func main() {
	var sum int64 = 0
	for y := 0; y < 500; y++ {
		for x := 0; x < 500; x++ {
			c_re := -2.0 + float64(x)*3.0/499.0
			c_im := -1.5 + float64(y)*3.0/499.0
			sum += int64(mandelbrot(c_re, c_im, 1000))
		}
	}
	fmt.Println(sum)
}

File: benchmarks/mandelbrot/mandelbrot.rs
fn mandelbrot(c_re: f64, c_im: f64, max_iter: i32) -> i32 {
    let mut z_re = 0.0f64;
    let mut z_im = 0.0f64;
    for i in 0..max_iter {
        let z_re_sq = z_re * z_re;
        let z_im_sq = z_im * z_im;
        if z_re_sq + z_im_sq > 4.0 {
            return i;
        }
        z_im = 2.0 * z_re * z_im + c_im;
        z_re = z_re_sq - z_im_sq + c_re;
    }
    max_iter
}

fn main() {
    let mut sum: i64 = 0;
    for y in 0..500i32 {
        for x in 0..500i32 {
            let c_re = -2.0 + (x as f64) * 3.0 / 499.0;
            let c_im = -1.5 + (y as f64) * 3.0 / 499.0;
            sum += mandelbrot(c_re, c_im, 1000) as i64;
        }
    }
    println!("{}", sum);
}
